File: QuadraticRegion.py
#!usr/bin/env python3

import math
from Facet import Template

def prep():
    global ball_colour, x_ball, y_ball, xq_lb, xq_ub, turtle_radius

    # bind fun function
    scene.button.onrelease(play)
    
    turtle.penup()
    # button text
    turtle.setposition(scene.button.xcor(), scene.button.ycor()-10)
    turtle.pendown()
    turtle.write("Move", align="center", font=("Arial",16,"bold"))

    # ball parameters
    diameter = 5
    ball_colour = "#ffff00"
    x_ball = 100 - scene.wnd_width/2; y_ball = 0

    # quadratic curve parameters
    xq_start = -200; yq_start = scene.wnd_height/3
    xq_min = 0; yq_min = -scene.wnd_height/3
    
    a = (yq_start - yq_min) / math.pow((xq_start - xq_min), 2)
    b = -2*a*xq_min
    c = yq_min + a*math.pow(xq_min, 2)

    # curve bounds
    discriminant = math.sqrt(b*b - 4*a*(c - y_ball))
    if a < 0: # a is -ve
        xq_lb = (-b + discriminant) / (2*a) # upper boundary
        xq_ub = (-b - discriminant) / (2*a) # lower boundary
    else: # a is +ve
        xq_lb = (-b - discriminant) / (2*a) # lower boundary
        xq_ub = (-b + discriminant) / (2*a) # upper boundary

    # draw quadratic curve
    xq = xq_start; yq = yq_start
    turtle.penup()
    turtle.setposition(xq, yq)
    turtle.pendown()
    while xq <= -xq_start:
        yq = a*xq*xq + b*xq + c
        turtle.setposition(xq, yq)
        xq += 1
    
    screen.delay(20)
    
    # transform turtle into a ball
    turtle.penup()
    turtle.setposition(x_ball, y_ball)
    turtle.setheading(0)
    turtle.shape("turtle")
    turtle.shapesize(diameter, diameter)
    turtle.color(ball_colour, ball_colour)
    
    turtle_radius = 10*turtle.shapesize()[1]

# fun function when button is clicked
# just moves turtle until it hits the right boundary
def play(x, y):
    global x_ball, y_ball, ball_colour, xq_lb, xq_ub, turtle_radius
    
    if x_ball < scene.wnd_width/2 - turtle_radius:
        # yellow outside the quadratic region
        ball_colour = "#ffff00"
        if (x_ball - turtle_radius <= xq_lb and x_ball + turtle_radius >= xq_lb) or\
           (x_ball - turtle_radius <= xq_ub and x_ball + turtle_radius >= xq_ub):
            # red on the quadratic outline
            ball_colour = "#ff0000"
        elif x_ball - turtle_radius >= xq_lb and x_ball + turtle_radius <= xq_ub:
            # green inside the quadratic region
            ball_colour = "#00ff00"
        turtle.color(ball_colour, ball_colour)
            
        # change ball's coordinates
        turtle.setposition(x_ball, y_ball)

        x_ball += 10
        screen.ontimer(play(0,0), 500)
    else:
        x_ball = 100 - scene.wnd_width/2; y_ball = 0
        screen.mainloop()

# Let fly
scene = Template();
turtle = scene.controlButtons()
screen = scene.screen

prep()

File: QuadraticRegion.js
var canvas = document.getElementById("quad_region");
var context = canvas.getContext("2d");
context.fillStyle = "#888888"; // color for our moving body(circle)

// coordinates for the ball(circle)
var x_ball = 50;
var y_ball = 150;
var radius = 40;
// draw the circle
context.beginPath();
context.arc(x_ball, y_ball, radius, 0, 2 * Math.PI);
context.fill();

//quadratic coordinates
var xq_start = 200;
var yq_start = 50;
var xq_min = 350;
var yq_min = 300;
var xq_stop = 500;

var a = (yq_start - yq_min) / Math.pow((xq_start - xq_min), 2);
var b = -2 * a * xq_min;
var c = yq_min + a * Math.pow(xq_min, 2);

//draw quadratic curve
var x = xq_start;
var y = yq_start;
for (; x < xq_stop; x += 1) {
    // redraw dot
    y = a * x * x + b * x + c;
    context.beginPath();
    context.arc(x, y, 1, 0, 2 * Math.PI);
    context.fill();
}

var discriminant = Math.sqrt(b * b - 4 * a * (c - y_ball));
var xq_lb;
var xq_ub;
if (a < 0) {// a is negative
    xq_lb = (-b + discriminant) / (2 * a); // upper boundary
    xq_ub = (-b - discriminant) / (2 * a); // lower boundary
} else {
    xq_lb = (-b - discriminant) / (2 * a); // lower boundary
    xq_ub = (-b + discriminant) / (2 * a); // upper boundary
}

var clr_obj;
function doGlide() {
    // condition for continuing motion
    if (x_ball <= 540) {
        context.clearRect(x_ball - radius - 10, y_ball - radius, 80, 80); // erase previous circle
        if ((x_ball - radius <= xq_lb && x_ball + radius >= xq_lb)
                || (x_ball - radius <= xq_ub && x_ball + radius >= xq_ub)) {
            context.fillStyle = "#ff0000"; // trespaaing color for our moving body(circle)
        } else if (x_ball - radius >= xq_lb && x_ball + radius <= xq_ub) {
            context.fillStyle = "#00ff00"; // zone color for our moving body(circle)
        } else {
            context.fillStyle = "#888888"; // out of zone color for our moving body(circle)
        }
        // redraw circle
        context.beginPath();
        context.arc(x_ball, y_ball, radius, 0, 2 * Math.PI);
        context.fill();

        x_ball += 10;
        clr_obj = setTimeout(function () { doGlide(); }, 100);
    }
}